documentation/src/main/asciidoc/topics/con_off_heap_memory.adoc

```diff
@@ -0,0 +1,66 @@
+However there are some trade-offs such as that heap dumps become worthless.
+
+Consider the following illustration of memory space for a JVM process where
+{brandname} is running:
+
+image::offheap.svg[]
+
+JVM heap memory, as shown in the preceding diagram, holds Java objects and other application data.
+The heap is divided into young and old generations that help keep referenced objects in memory.
+The Garbage Collector (GC) process reclaims space from unreachable objects, running more frequently on the young generation memory pool.
+
+When {brandname} stores cache entries in JVM heap memory, GC runs can take longer to complete as you start adding data to your caches.
+Because GC is a fairly intensive process, longer and more frequent runs can degrade application performance.
+
+Using off heap memory configures {brandname} to store data in native memory, or available system memory that is outside JVM memory management.
+Depending on your JVM implementation, native memory is also allocated for thread stacks, Just-in-Time (JIT) compiler code, and so on.
+As an example, the preceding diagram shows two memory pools that are allocated from native memory:
+
+* Metaspace holds class metadata.
+* Code Cache holds generated native code and internal structures.
+
+Taking the diagram as an example, when you use off heap storage {brandname} adds data to the Code Cache memory pool.
+
+When you add an entry to a cache that uses off heap storage, {brandname} hashes the serialized `byte[]` for the key into buckets that are similar to a standard Java `HashMap`.
+Buckets include address pointers that {brandname} use to locate entries in off heap memory.
+
+The following diagram shows a set of keys with names, the hash for each key and bucket array of address pointers, as well as the entries with the name and phone number:
+
+image::offheap_hashmap.svg[]
+
+In cases where key hashes collide, {brandname} chains entries.
+As in the preceding diagram, if the William Clay and Luke Cage keys have the same hash, then the first entry added to the cache is the first element in the bucket.
+
+.Memory overhead
+
+Memory overhead is the additional memory that {brandname} uses to store entries.
+{brandname} uses 25 bytes for each entry entry in the cache.
+
+Using eviction to create a bounded off heap data container increases memory overhead by 36 bytes.
+This happens because {brandname} creates an additional linked list to track entries in the cache and perform eviction.
+
+However, compared with storing entries in the JVM heap, a total memory overhead of 61 bytes per entry is significantly lower.
+
+.Data consistency
+
+{brandname} uses an array of `ReadWriteLock` instances to protect off heap address spaces.
+The number of `ReadWriteLock` instances is twice the number of cores and then rounded to the nearest power of two.
+This ensures that there is an even distribution of `ReadWriteLock` instances to prevent read operations from blocking each other.
```

Comment:
The distribution of read locks is not so they wouldn't block each other as a the read portion of a read write lock will not block another read. The distribution is to prevent reads from being blocked by a write lock.